dask/array/slicing.py

```diff
@@ -1031,6 +1031,16 @@ def slice_with_bool_dask_array(x, index):
     ]
 
     if len(index) == 1 and index[0].ndim == x.ndim:
+        if not np.isnan(x.shape).any() and not np.isnan(index[0].shape).any():
+            x = x.ravel()
+            index = tuple(i.ravel() for i in index)
+        elif x.ndim > 1:
+            warnings.warn(
+                "When slicing a dask array of unknown chunks with a boolean mask "
+                "dask array, the output array may have a different ordering "
+                "compared to the equivalent NumPy operation.",
+                stacklevel=3,
```

Comment:
This seems unsafe, should we eventually error here instead (a warning is certainly better now than silently being incorrect before)? If we can't maintain order reductions are still fine, but other code may break.

Reply:
I agree this isn't ideal. I'm slightly in favor of just raising a warning in order to not break any existing user code. The warning will at least inform them that the existing behavior doesn't match up with NumPy. Do you have other thoughts / feel strongly about warning vs. erroring?

Reply:
I suppose another option would be to have a config flag that users set to get NumPy ordering. We could then warn that this only works correctly for *some* cases (like arrays of known size). This would mean users have to opt-in to this behavior and they are warned in advance of the consequences.

Reply:
I'm against adding another config option - we're already swimming in a sea of config options, I think we should be opinionated in places like this. If we really do need to make this configurable, then it should be with a keyword argument to allow the non-compatible behavior (`strict_ordering=False`?), with the default to numpy compatible/errors (the error could point to this flag). I think the number of users a breaking change like this would affect would be small, and the risk of incorrect output is worse IMO than breaking their code. Thoughts?

Reply:
I'd also prefer not to add a configuration option for this. Rather, treat the ordering mismatch as a bug and either warn or raise. 

> I think the number of users a breaking change like this would affect would be small, and the risk of incorrect output is worse IMO than breaking their code

I'm not sure how many would be affected by erroring here. A good compromise might be to warn now and elevate to an error after we've issued a release with the warning included.

Reply:
> I'm not sure how many would be affected by erroring here. A good compromise might be to warn now and elevate to an error after we've issued a release with the warning included.

A warning that's elevated to an error later would be fine with me.

Reply:
Just added a note about the warning changing to an error in a future release

Reply:
Are we happy with `UserWarning` or is there another warning type we should be using here?

Reply:
We use UserWarning for most things in dask, fine by me.